### PR TITLE
Check resource name on included serializer in to_internal_value

### PR DESCRIPTION
**Pull request by czosel**
Fixes https://github.com/django-json-api/django-rest-framework-json-api/issues/302 and #303

### CODE REVIEW

**Comment by czosel** [comment]
@mblayman Did you get a chance to take a look at this one yet? Please let me know if there's anything that needs more clarification or work. 
Please don't feel pressured by my question - it's just that since we're starting to depend on this fix more and more internally I decided to ask, to get a feeling for our chances of this being merged in the future :smile: 

PS: Thank you for taking over the maintenance for this repo, you're doing a great job! :+1:

**Comment by mblayman** [comment]
@czosel this PR pre-dated my start as a maintainer role so I haven't really looked at it closely. I'll try to get some time to take a look at it soon.

Thanks for the contribution!

**Comment by czosel** [comment]
@mblayman Awesome, thanks for your quick reply!

File: example/tests/integration/test_model_resource_name.py [hunk 1]
```diff
@@ -61,29 +80,43 @@ def test_resource_name_precendence(self, client):
...
+        response = client.post(reverse("comment-list"),
+                               dump_json(self.create_data),
+                               content_type='application/vnd.api+json')
+
+        assert response.status_code == status.HTTP_201_CREATED
+
+    def test_serializer_resource_name_create(self, client, monkeypatch):
+        monkeypatch.setattr(serializers.CommentSerializer.Meta, 'resource_name', 'renamed_comments', False)
+        monkeypatch.setattr(serializers.EntrySerializer.Meta, 'resource_name', 'renamed_entries', False)
+        self.create_data['data']['type'] = 'renamed_comments'
```
Comment (mblayman) [inline example/tests/integration/test_model_resource_name.py]: I think you should probably deepcopy `self.create_data`. Otherwise, I think you're modifying the class state for all tests.

File: example/tests/integration/test_model_resource_name.py [hunk 2]
```diff
@@ -61,29 +80,43 @@ def test_resource_name_precendence(self, client):
...
+
+        response = client.post(reverse("comment-list"),
+                               dump_json(self.create_data),
+                               content_type='application/vnd.api+json')
+
+        assert response.status_code == status.HTTP_201_CREATED
+
     def teardown_method(self, method):
         models.Comment.__bases__ = (models.Comment.__bases__[0],)
-        try:
```
Comment (mblayman) [inline example/tests/integration/test_model_resource_name.py]: I'm not super familiar with these tests. Would you mind explaining why this tear down code is not needed?

File: rest_framework_json_api/relations.py [hunk 3]
```diff
@@ -142,19 +148,44 @@ def to_representation(self, value):
...
+    def get_root_serializer(self):
+        if hasattr(self.parent, 'parent') and self.is_serializer(self.parent.parent):
+            return self.parent.parent
+        elif self.is_serializer(self.parent):
+            return self.parent
+
+        return None
+
+    def is_serializer(self, candidate):
+        return hasattr(candidate, 'included_serializers')
```
Comment (mblayman) [inline rest_framework_json_api/relations.py]: Is this a true test of whether something is a serializer? A serializer could be somethign that donesn't have `included_serializers`, right? Am I mistaken?

**Comment by mblayman** [comment]
@czosel This stuff looks pretty good to me. Thanks for your effort. I have a couple of questions which I added as comments.

Could I also request that you merge the latest `develop` branch, add your name to the AUTHORS file (since this is a community project and I think everyone involved should get some credit), and add an entry to the CHANGELOG for v2.3.0?

Also, I've heard that features that aren't documented don't exist. Could we find an appropriate place to document this functionality so others might benefit from your work?

Thanks!

File: example/tests/integration/test_model_resource_name.py (hunk 2 above)
Comment (czosel) [inline example/tests/integration/test_model_resource_name.py]: No problem!
Instead of setting the `resource_name` properties directly, i'm using `monkeypatch.setattr`, which will remember the previous state of the attribute and undo the change after the test ran. For more info on monkeypatch see this [blog post](https://holgerkrekel.net/2009/03/03/monkeypatching-in-unit-tests-done-right/) or the [docs](https://docs.pytest.org/en/latest/monkeypatch.html).

File: example/tests/integration/test_model_resource_name.py (hunk 1 above)
Comment (czosel) [inline example/tests/integration/test_model_resource_name.py]: Agreed!

File: rest_framework_json_api/relations.py (hunk 3 above)
Comment (czosel) [inline rest_framework_json_api/relations.py]: You're right - i just verified that `isinstance(candidate, Serializer)` also works in this case, which is certainly cleaner.

**Comment by czosel** [comment]
@mblayman Thanks for your review! I just pushed some minor changes and hope to have answered your questions. The only thing i didn't do was change the documentation, because i think the ["setting the resource name chapter"](http://django-rest-framework-json-api.readthedocs.io/en/stable/usage.html?highlight=resource_name#setting-the-resource-name) in the docs pretty much says it all - this PR is basically just fixing this feature for two previously broken cases (nested serializers, POST requests).

**Comment by mblayman** [comment]
👍 Thanks for your patience, @czosel! Sorry it took 6 months to get this work merged.

**Comment by czosel** [comment]
@mblayman No worries, thanks for merging it! :+1:
